File: Java/src/parser/util/ParserUtils.java
package parser.util;

import org.apache.commons.lang.StringUtils;
import parser.ast.ASTNode;
import parser.ast.Factor;

import java.util.ArrayList;
import java.util.LinkedList;

/**
 * Author : lihuichuan
 * Time   : 2020/12/7
 */
public class ParserUtils {

    public static String toPostfixExpression(ASTNode node) {

        if (node instanceof Factor) {
            return node.getLexeme().getValue();
        }

        ArrayList prts = new ArrayList<String>();
        for (ASTNode child : node.getChildren()) {
            prts.add(toPostfixExpression(child));
        }
        String lexemeStr = node.getLexeme() != null ? node.getLexeme().getValue() : "";
        if (lexemeStr.length() > 0) {
            return StringUtils.join(prts, " ") + " " + lexemeStr;
        } else {
            return StringUtils.join(prts, " ");
        }

    }

    public static String toBFSString(ASTNode root, int max) {
        LinkedList<ASTNode> queue = new LinkedList<ASTNode>();
        ArrayList res = new ArrayList<String>();

        queue.add(root);
        int c = 0;
        while (queue.size() > 0 && c++ < max) {
            ASTNode node = queue.poll();
            res.add(node.getLabel());
            for (ASTNode child : node.getChildren()) {
                queue.add(child);
            }
        }
        return StringUtils.join(res, " ");
    }
}

File: JS/src/parser/util/ParserUtils.js
const Factor = require('../ast/Factor')
const ASTNodeTypes = require('../ast/ASTNodeTypes')
const LinkedList = require('linkedlist')

class ParserUtils {
    static toPostfixExpression(node) {
        if (node instanceof Factor) {
            return node.getLexeme().getValue()
        }

        const prts = []
        for (const child of node.getChildren()) {
            prts.push(ParserUtils.toPostfixExpression(child));
        }
        const lexemeStr = node.getLexeme() != null ? node.getLexeme().getValue() : "";
        if (lexemeStr.length > 0) {
            return prts.join(" ") + " " + lexemeStr
        } else {
            return prts.join(" ")
        }
    }

    static toBFSString(root,max){
        const queue = new LinkedList()
        const list = []
        queue.push(root)

        let c = 0
        while(queue.length > 0 && c++ < max) {
            const node = queue.shift()
            list.push(node.getLabel())
            for(const child of node.getChildren()) {
                queue.push(child);
            }
        }
        return list.join(" ")
    }
}
module.exports = ParserUtils